src/onnx/onnx.cpp

```diff
@@ -1697,10 +1713,16 @@ struct onnx_parser
             op::lstm{hidden_size, vec_actv_funcs, dirct, clip, input_forget}, std::move(args));
 
         // second output for last lstm output
-        auto last_output = prog.add_instruction(op::rnn_last_output{}, hidden_states);
+        std::vector<instruction_ref> vec_args;
+        vec_args.push_back(hidden_states);
+        if(seq_lens != prog.end())
+        {
+            vec_args.push_back(seq_lens);
+        }
+        auto last_output = prog.add_instruction(op::rnn_last_hs_output{dirct}, vec_args);
 
         // third output for last cell output
-        auto last_cell_output = prog.add_instruction(op::lstm_last_cell_output{}, hidden_states);
+        auto last_cell_output = prog.add_instruction(op::lstm_last_cell_output{dirct}, vec_args);
```

Comment:
Why does `lstm_last_cell_output` need to take direction?

Reply:
The cell output is of shape [num_direction, batch_size, hidden_size]. If the any sequence len is less than max_sequence_len, we need the direction information to know whether to retrieve the last element from all_cell_outputs (forward direction) or the first element from the all_cell_outputs (reverse direction)

Reply:
But wont the shape be the same for the last element and first element?

Reply:
It is fine for the shape computation. For the case of different input sequence length, the algorithm used to compute the last output in forward and reverse direction is different. Please see the following:
       visit_all(result, args[0])([&](auto output, auto input) {
            args[1].visit([&](auto seq_lens) {
                par_for(output_shape.elements(), [&](auto i) {
                    auto idx = out_comp_s.multi(i);
                    auto b   = idx[2];
                    if(direction == rnn_direction::reverse or idx[1] == 1)
                    {
                        idx[0] = 0;
                    }
                    else
                    {
                        idx[0] = seq_lens[b] - 1;
                    }
                    output[i] = input(idx.begin(), idx.end());
                });
            });
        });
so we need the direction in computing the last output

Reply:
Yes, but you only need the direction for the actual operator. This shouldn't be there for the placeholder. The rewrite_rnn will change it over to use the operator with direction.

Reply:
fixed